### ISSUE TITLE
[Feature Request] Add `to`, `cpu`, and `cuda` method to optimizer

### ISSUE THREAD

**Issue by jeongukjae** [pytorch/pytorch#41839]
## 🚀 Feature

```python
model =     # initialize
optimizer = # initialize

model.cuda()

optimizer.load_state_dict(optimizer_checkpoint)  # if checkpoint is not on gpu memory
optimizer.cuda()                                 # <<<<<< this code can be helpful

# and training codes...
loss.backward()
optimizer.step()
```

I think there is no appropriate way to send optimizer states to gpu memory currently.

## Motivation

When training should be saved and resumed, I packed all state dict to one dictionary object like `{"model": model.module.state_dict(), "optimizer": optimizer.state_dict(), "lr_scheduler": scheduler.state_dict()}` and save it. Due to the limit of GPU memory, when resuming training, I always load state dict using `map_location=torch.device('cpu')`.

At that situation, I have to cast all optimizer state to gpu manually like below code(From issue https://github.com/pytorch/pytorch/issues/2830).

```python
for state in optimizer.state.values():
    for k, v in state.items():
        if torch.is_tensor(v):
            state[k] = v.cuda()
```

## Pitch

Implement `to`, `cpu`, and `cuda` method to `optim.Optimizer` like `nn.Module`.

## Additional Context

Maybe related to #2830

**Comment by mrshenli** [comment]
Self-assigned to reserve this for bootcamp.

**Comment by vincentqb** [comment]
Good idea! One question to think about from [comment](https://github.com/pytorch/pytorch/issues/2830#issuecomment-336183179), if the model isn't on cuda device, and the optimizer is moved, would the optimizer silently not update the parameter? I have not tested this :)

**Comment by mrshenli** [comment]
un-assigned myself, this is no longer reserved for bootcamp. Please feel free to grab if you are interested.

**Comment by Kadam-Tushar** [comment]
is someone working on this? If nobody is working on this I wish to work on this.
